src/Microsoft.Health.Fhir.TemplateManagement/Client/OrasClient.cs

```diff
@@ -26,38 +29,72 @@ public OrasClient(string imageReference)
+            }
+
+            // Oras will create output folder if not existed.
+            // Therefore, the output folder should exist if pull succeed.
+            if (!Directory.Exists(outputFolder))
+            {
+                throw new OCIClientException(TemplateManagementErrorCode.OrasProcessFailed, "Pull image failed or image is empty.");
+            }
+
+            try
+            {
+                string manifestContent = LoadManifestContentFromCache(digest);
+                var manifest = JsonConvert.DeserializeObject<ManifestWrapper>(manifestContent);
+                File.WriteAllText(Path.Combine(outputFolder, Constants.ManifestFileName), LoadManifestContentFromCache(digest));
+                return manifest;
+            }
+            catch (Exception ex)
+            {
+                throw new OCIClientException(TemplateManagementErrorCode.OrasCacheFailed, "Read manifest from oras cache failed.", ex);
+            }
         }
 
-        public async Task PushImageAsync(string inputFolder, List<string> filePathList)
+        public async Task PushImageAsync(string inputFolder)
         {
+            if (!Directory.Exists(inputFolder))
+            {
+                throw new OverlayException(TemplateManagementErrorCode.ImageLayersNotFound, "Image folder not found.");
+            }
+
             string argument = string.Empty;
             string command = $"push \"{_imageReference}\"";
+            var filePathesToPush = Directory.EnumerateFiles(inputFolder, "*.tar.gz", SearchOption.AllDirectories);
 
-            // In order to remove image's directory prefix. (e.g. "layers/layer1.tar.gz" --> "layer1.tar.gz"
+            // In order to remove image's directory prefix. (e.g. "layers/layer1" --> "layer1"
             // Change oras working folder to inputFolder
-            foreach (var filePath in filePathList)
+            foreach (var filePath in filePathesToPush)
             {
-                if (!File.Exists(Path.Combine(inputFolder, filePath)))
-                {
-                    throw new OverlayException(TemplateManagementErrorCode.ImageLayersNotFound, "Image layer not found");
-                }
-
-                argument += $" \"{filePath}\"";
+                argument += $" \"{Path.GetRelativePath(inputFolder, filePath)}\"";
             }
 
             if (string.IsNullOrEmpty(argument))
             {
-                throw new OverlayException(TemplateManagementErrorCode.ImageLayersNotFound, "No file for push.");
+                throw new OverlayException(TemplateManagementErrorCode.ImageLayersNotFound, "No file to push.");
             }
 
             await OrasExecutionAsync(string.Concat(command, argument), inputFolder);
         }
 
-        public static async Task OrasExecutionAsync(string command, string orasWorkingDirectory)
+        public static async Task<string> OrasExecutionAsync(string command, string orasWorkingDirectory)
```

Comment:
need to be public?

Reply:
After offline discussion, the function will be used by tests.